### ISSUE TITLE
How to use Flash Attention 2 with huggingface models ?

### ISSUE THREAD

**Issue by MohamedAliRashad** [Dao-AILab/flash-attention#320]
How to use Flash Attention 2 with huggingface models ?

**Comment by tridao** [comment]
I don't know of a good way, except to rewrite the model code to call FlashAttention functions to compute the scaled dot-product attention softmax(Q @ K^T) @ V. 
Maybe there's some automated way to monkey patch, but I'm not familiar.

**Comment by Narsil** [comment]
depends on the model, for generative models you can check out : https://github.com/huggingface/text-generation-inference

**Comment by learning-chip** [comment]
For FlashAttention1, [`optimum.bettertransformer`](https://huggingface.co/docs/optimum/bettertransformer/overview) can be used to transform HF models to use `scaled_dot_product_attention` in PT2.0, which then calls to FlashAttention-1. If FlashAttention-2 is also made available for `scaled_dot_product_attention`, then I think it can be used in the same way?

**Comment by fxmarty** [comment]
Yes indeed @learning-chip. The big issue with PyTorch SDPA & the transformers implementation is the need of a padding mask for batch size > 1, which makes things a bit hard to be compatible with flash. Hopefully we an improve that in the near future.
